src/algorithms/mul_redc.rs

```diff
@@ -1,70 +1,289 @@
-            }
-            if r > m {
-                break;
-            }
+
+    // Compute reduced product.
+    reduce1_carry(result, modulus, carry)
+}
+
+/// Computes a^2 * 2^(-BITS) mod modulus
+///
+/// Requires that `inv` is the inverse of `-modulus[0]` modulo `2^64`.
+/// Requires that `a` is less than `modulus`.
+#[inline]
+#[must_use]
+#[allow(clippy::cast_possible_truncation)]
+pub fn square_redc<const N: usize>(a: [u64; N], modulus: [u64; N], inv: u64) -> [u64; N] {
+    debug_assert_eq!(inv.wrapping_mul(modulus[0]), u64::MAX);
+    debug_assert!(less_than(a, modulus));
+
+    let mut result = [0; N];
+    let mut carry_outer = 0;
+    for i in 0..N {
+        // Add limb product
+        let (value, mut carry_lo) = carrying_mul_add(a[i], a[i], result[i], 0);
+        let mut carry_hi = false;
+        result[i] = value;
+        for j in (i + 1)..N {
+            let (value, next_carry_lo, next_carry_hi) =
+                carrying_double_mul_add(a[i], a[j], result[j], carry_lo, carry_hi);
+            result[j] = value;
+            carry_lo = next_carry_lo;
+            carry_hi = next_carry_hi;
+        }
+
+        // Add m times modulus to result and shift one limb
+        let m = result[0].wrapping_mul(inv);
+        let (value, mut carry) = carrying_mul_add(m, modulus[0], result[0], 0);
+        debug_assert_eq!(value, 0);
+        for j in 1..N {
+            let (value, next_carry) = carrying_mul_add(modulus[j], m, result[j], carry);
+            result[j - 1] = value;
+            carry = next_carry;
+        }
+
+        // Add carries
+        if modulus[N - 1] >= 0x3fff_ffff_ffff_ffff {
+            let wide = (carry_outer as u128)
+                .wrapping_add(carry_lo as u128)
+                .wrapping_add((carry_hi as u128) << 64)
+                .wrapping_add(carry as u128);
+            result[N - 1] = wide as u64;
+
+            // Note carry_outer can be {0, 1, 2}.
+            carry_outer = (wide >> 64) as u64;
+            debug_assert!(carry_outer <= 2);
+        } else {
+            // `carry_outer` and `carry_high` are always zero.
+            debug_assert!(!carry_hi);
+            debug_assert_eq!(carry_outer, 0);
+            let (value, carry) = carry_lo.overflowing_add(carry);
+            debug_assert!(!carry);
+            result[N - 1] = value;
         }
     }
-    if reduce {
-        // REFACTOR: Create sub routine
-        let mut carry = 0;
-        #[allow(clippy::cast_possible_truncation)] // Intentional
-        #[allow(clippy::cast_sign_loss)] // Intentional
-        for (r, m) in zip(result.iter_mut(), m.iter().copied()) {
-            carry += i128::from(*r) - i128::from(m);
-            *r = carry as u64;
-            carry >>= 64; // Sign extending shift
+
+    // Compute reduced product.
+    debug_assert!(carry_outer <= 1);
+    reduce1_carry(result, modulus, carry_outer > 0)
+}
+
+#[inline]
+#[must_use]
+fn less_than<const N: usize>(lhs: [u64; N], rhs: [u64; N]) -> bool {
```

Comment:
Can use `algorithms::cmp` https://github.com/recmo/uint/blob/b91e6f710e42b893e695cae43ffab3d65740d881/src/algorithms/mod.rs#L94

Reply:
also, what benefit would we get from monomorphizing here? would it be more/less efficient to do `for i in 0..N` instead of `zip`? I need to develop better intuition for this 🤔

Reply:
I've found concrete performance to be better with `<const N: usize>(x: [u64; N])` rather than `(x: &[u64])`. In particular for Montgomery multiplication we benefit a lot from the compiler unrolling the loop when `N` is small and inlining the `modulus` when it is a compile time constant (which also eliminates the modulus-dependent branches).

Reply:
I've found concrete performance to be better with `<const N: usize>(x: [u64; N])` rather than `(x: &[u64])`. In particular for Montgomery multiplication we benefit a lot from the compiler unrolling the loop when `N` is small and inlining the `modulus` when it is a compile time constant (which also eliminates the modulus-dependent branches).

Reply:
In this case `less_than` was no longer used in production code, so no reason to keep it.